### PR TITLE
Continuous Nebulosity Masking

### PR DESCRIPTION
**Pull request by andrew-saydjari**
The purpose of this PR is mainly to merge in the continuous nebulosity masking capabilities that have been developed by me for DECaPS2 in order to make them available for @schlafly processing of WISE data. In addition, minor functionalities have been added (listed below).

All stable changes have been implemented to leave the default behavior of crowdsource prior to this PR unchanged (though the user interface may have changed ever so slightly). To change the photometric solutions requires throwing new flags and settings which are now accessible. I hope this promotes back compatibility and eases validation. In many cases, in `verbose` mode, the code will currently warn that the behavior has changed.

Stable Changes:
- `psvalsharpcutfac` and `psfsharpsat` access at the `fit_im` level for diagnostic tuning (instead of hard coding)
- the probability of each class from the nebulosity NN at the central pixel of each star is exported to the catalogue. These are now accessibly as `pN`, `pL`, `pR`, and `pE`. In the same change, we also store whether or not the star is in a nebulosity masked region as `neb_mask`, but this is probably redundant with the `dq` bits which are set to nebulous and could be removed.
- the initial level of deblending is available at the `fit_im` level for diagnostic tuning as `blendthreshu` and the iteration at which the `fit_im` becomes more aggressive, `blendthresh` = 0.2, is available as a command line option `titer_thresh`.
- the `passno` (the iteration) at which a source is found during the `fit_im` loop is now saved to the catalogue
- the extension is printed at each `fit_im` step to help clarify cmdline output in the ccd-level parallelization case.
- decam file names are now read in from the cmd line by survey, date, filtf, and vers which then defines the ooi, oow, ood, mod, and cat file names rather than have to specify all of the later all by hand.
- S7 correction now occurs before nebulosity masking (image analysis by the NN) because the sub 512x512 shifts introduced by gen_prob can cause the test images to include the boundary between amplifiers.
- the decision boundary for nebulous masking is now a linear cut in the probability simplex set by the `alpha` and `gamma` hyperparameters instead of a max in probability space.
- guppy memory profiling has been added to the functionality enabled by the `--profile` cmd line option
- the masking of bright stars (which was hurting DECaPS2 residuals) can now be easily switched on/off by the `--bmask_off` cmd-line option
- `outdir` is now split into `outdirc` and `outdirm` for cat and model respectively
- `ccdlist` cmd-line option lets the user specify a list of ccds to do to enable ccd targeted test runs (when run in resume mode, only the incomplete subset of `cddlist` will be run)
- cmdline access to `pixsz` for the `decam_psf` is enabled
- the entire bool nebulosity_mask is saved as part of the model file (i4 is cheap)
- if continuous masking is used, a ternary diagram of P, L, and R+E is made at low resolution in the probability simplex and saved to the model file
- nproc limits on the total number of ccds to be run is changed from > to >= so that nproc of 1 means 1 ccd will be run
- a ccd-level parallism version of the decam process_image function was made to enable fast full exposure runs. This is used with the `--parallel` cmd-line option which switches to `process_image_p`. Here we use pqdm to distribute and collect the ccd jobs and get a nice progress bar as an added bonus.
- a `galaxy_mask` module was added which is an almost exact translation of IDL code from @ameisner into python. This reads in a hyperleada list of galaxies, applies some by hand modifications to that list (from @ameisner), and then compares the positions of those galaxies to the pixels in each ccd during processing. In case of an overlap, a new extrabit 2**24 for "galaxy" is set and the nodeblend_maskbit 2**30 is set in that region. When the WCSCAL fails so that we cannot accurately compare the position of galaxies to the ccd, the galaxy_masking is not performed and is skipped verbosely.
- tensorflow warnings about optimizing NN performance have been silenced forcibly
- `subimages` in the `nebulosity_mask` module has been modified to take a `shiftx` and `shifty` parameter in order to define shifted grids of our fixed test/train image size of 512 x 512 relative to the image dimensions
- the usage of the old `gen_mask` should be replaced by `gen_prob`. This is currently hard coded to use 128 pixel subshifts of the test/train images, perform a weighted average of the class probailities (weighted in favor of the nebulosity class), and then smooth the resulting probability images by a 128x128 sigma gaussian filter. A decision can be made off of these probability images better than off the `gen_mask` outputs (see decap_proc.py notes for more). This functionality is turned on bythe `contmask` cmd-line option.
- `plot_psf_fits` now takes the extension name and iteration number as arguments in order to save images of the residuals of the top 100 "brightest good" stars used for the psf fitting at each round of `fit_im`. This behavior (which is for diagnostics only) can be enabled by the --plot_on cmd line flag.

Experimental changes represent working code that is under development but is not yet fully integrated or ready for general use.

Experimental Changes:
- `bin_weights` switches the least squares fit from inverse variance weighting to essentially no weighting, excluding only bad pixels from the fit. No effort has been made to modify `dflux` or `flux_lbs`/`dflux_lbs` to agree with this weighting scheme.
- `w_cutoff` was introducted (cmd-line access) as a floor below which to set all weights to zero. This is not really necessary when the dq is used properly as it is already in decam_proc.py, but could be useful for other photometeric products and so will be retained.

Reorganization:
- example folder added to hold wise_proc_example.sh, decam_cmds.txt (which holds some examples of cmd line submission with the old and new command line options for decam), and makecat.py (from DECaPS1 run).
-  requirements.txt was added in the hopes that we could define a reproducible and stable package environment to ease reproducibility on new clusters and for new users. This includes the new packages required for this PR (python-ternary, pydl, pqdm, colorcet (optional), and guppy3)

Thoughts on changes needed for application to WISE:
- a wise_gen_prob function needs to be written, by analogy to the current gen_prob and wise_gen_mask functions, and the decision boundary (if it changes from a simple argmax) in terms of `alpha` and `gamma` should be validated.
- the model file outputs for contmask (pN, pL, pR, pE for each star and the ternary diagram) would need to change slightly to account for the fact that WISE seems to have only 3 classes in its NN model.

I will not plan to do anything to address those issues for WISE unless asked,  but let me know if you want help with those changes.

### CODE REVIEW

**Comment by andrew-saydjari** [comment]
This PR is now revalidated after the last up and back of changes. The following bugs were fixed:
- fits name pattern for storing pN, pL, pR, pE
- ordering of pN, pL, pR, and pE corrected
- quantization level was too low for saving the pN, pL, pR, pE subsampled images, upped to 2 and downsampled 8x8 (1.2 MB increase in the mod compared to usual ~11 MB)
- caught nebprob called before assignment when -no-diffuse-mask is on

Validations:
- galaxy masking after converting to astropy is correct (concerned about arcmin/arcsec conversion factors of 60 flying around)
- stellar catalogue position and fluxes identical
- nebulosity probabilities at stellar centers changed at the 0.1% level in regions with large probability gradient. Also fixed a small difference in the default bc between the new scipy.ndimage and old skimage gaussian blur to always use a constant edge extrapolation. The 0.1% level changes appear to result from the fact that skimage converts image data to float64 before doing the gaussian blur (which is just a wrapped version of scipy.ndimage) whereas the new scipy.ndimage implementation will just process our inputs as float32s. 
- profiling with one processor works like usual and showed we are still at our expected ~460 MB memory/process. With a process pool, profiling hits the pool as expected and is less that helpful but runs to completion correctly.
- a variety of different command line option launches (including the new psf plotting/saving) were used to walk most if not all of the code paths.

I view this as ready to go for the unWISE run and pretty much the version I will use for DECaPS2 (except for the synthetic injection pipeline). While a lot of time was spent on changing this PR, the end of the day improvements are much cleaner/shorter parallel code (that also saves after every ccd, so much better intermediate checkpointing), fewer dependencies, and downsampled prb images instead of histograms to maintain more information in the mod outputs which is probably worth it in the end!

**Comment by schlafly** [comment]
Awesome, looks great @andrew-saydjari, merging.
